casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -75,10 +103,125 @@ object EstimatorHelper {
+    extractJoinedChannels(eventGroup1).intersect(allChannels(eventGroup2)).nonEmpty ||
+    extractJoinedChannels(eventGroup2).intersect(allChannels(eventGroup1)).nonEmpty ||
+    containConflictingEvents(eventGroup1, eventGroup2).nonEmpty
+  }
+
+  trait IsConflict
+  final case class NonConflict(leftEvents: EventGroup, rightEvents: EventGroup) extends IsConflict
+  final case class Conflict(
+      leftEvents: EventGroup,
+      rightEvents: EventGroup,
+      conflictSet: Set[Blake2b256Hash]
+  ) extends IsConflict
+
+  def isConflict[F[_]: Sync, C, P, A, K](
+      historyRepo: HistoryRepository[F, C, P, A, K],
+      baseState: Blake2b256Hash,
+      leftEvents: List[Event],
+      rightEvents: List[Event]
+  )(implicit sc: Serialize[C]): F[IsConflict] = {
+    val leftEventGroup  = extractEventGroup(leftEvents)
+    val rightEventGroup = extractEventGroup(rightEvents)
+    val conflictJoinInLeft =
+      extractJoinedChannels(leftEventGroup).intersect(produceChannels(rightEventGroup))
+    val conflictJoinInRight =
+      extractJoinedChannels(rightEventGroup).intersect(produceChannels(leftEventGroup))
+    val otherConflict          = containConflictingEvents(leftEventGroup, rightEventGroup)
+    val normalConflictChannels = conflictJoinInLeft ++ conflictJoinInRight ++ otherConflict
+    val nonconflictRightProduceChannels =
+      rightEventGroup.produces.filter(p => !normalConflictChannels.contains(p.channelsHash))
+    for {
+      rightJoins <- nonconflictRightProduceChannels.toList.traverse { produce =>
+                     for {
+                       joins <- historyRepo.getJoinsFromChannelHash(
+                                 baseState,
+                                 produce.channelsHash
+                               )
+                       joinM = joins.filter(_.length > 1)
+                     } yield (produce, joinM)
+                   }
+      leftProduceChannel = leftEventGroup.produces.map(_.channelsHash).toSet
+      conflictJoinChannels = rightJoins
+        .filter {
+          case (produce, joins) => {
+            val joinsChannelHashes  = joins.map(_.map(StableHashProvider.hash(_)(sc))).flatten
+            val joinsWithoutProduce = joinsChannelHashes diff Seq(produce.channelsHash)
+            joinsWithoutProduce.exists(p => leftProduceChannel.contains(p))
+          }
+        }
+        .map(_._1.channelsHash)
+        .toSet
+    } yield
+      if (normalConflictChannels.isEmpty && conflictJoinChannels.isEmpty) {
+        NonConflict(leftEventGroup, rightEventGroup)
+      } else {
+        Conflict(
+          leftEventGroup,
+          rightEventGroup,
+          normalConflictChannels ++ conflictJoinChannels
+        )
+      }
+  }
+
+  final case class MergeChanges(
+      validDeploys: List[ProcessedDeploy],
+      rejectedDeploys: List[ProcessedDeploy],
+      validEventLogs: Seq[RSpaceEvent]
+  )
+
+  def isDeploysConflict[F[_]: Sync, C, P, A, K](
```

Comment:
This method deserves description, as its important user facing one. Also name is misleading, as it tells that Boolean is returned. From what I see in the code its more like `computeMergeChanges`. 
Also - does `left` here corresponds to `main` and `right` to `merging`? So basically, return deploys will be subset of `right`.

Reply:
Also looks like this should be a method or extension for HistoryRepository?

Reply:
> Also - does left here corresponds to main and right to merging? So basically, return deploys will be subset of right.

 Yes.

I agree with you the this should a extension. EstimatorHelp is not a good place to hold  the codes.

Reply:
Already refactor a little bit.